essay/swar/latency.md

```diff
@@ -149,32 +173,45 @@ multiply128_to128_bits(unsigned __int128, unsigned __int128):           # @multi
+```
+{     A    |   B    |    C    |    D   }:  we want A+B+C+D.
+    Let's say we have *scalar* multiplication by very large integers and we can "cast" all lanes into a single scalar:
+    Then, we can craft a number of
+{     1    |    1   |    1     |    1  }: a scalar that encodes the number 1 every N bits where N is the lane width.
+Let's use the typical multiplication algorithm:
+     A          B        C         D +
+     B          C        D         0 +
+     C          D        0         0 +
+     D          0        0         0 =
+A+B+C+D       B+C+D     C+D        D
+The summation is in the most significant lane.
+```
+The issue with emulating missing SIMD horizontal operations is that you have to combine simple operations that diminish the gain factor of parallelism and require sophisticated mathematics, or what other people would call tricks.
+
+Up to when AVX and AVX2 were relatively new, changing the "type" of value had a non negligible penalty.  Because the execution units for FP and Integers were different.  This changed, nowadays they use the same execution units, as evidenced in the port allocations for the microarchitectures.  Apparently SIMD ALUs could be made that worke well with both scalar and SIMD values; but I conjecture that "something has got to give", I see that operations on 64 bit "scalar" integers and 512 bit `ZMM` values with internal structure of lanes are so different that "one type fits all" execution units would be less effective.  We can foresee ever wider SIMD computation, ARM's design has an extension that would allow immense widths of 2048 bits of SIMD, it is called "Scalable Vector Extensions", RISC-V has widening baked into the very design, they call it "Vector Length Agnostic" with its acronym VLA, so that you program in a way that is not specific to a width... Just thinking about the extreme challenge of merely propagating signals in buses of 2048, that the wider they are they also have to be longer, that the speed of signal propagation, the speed of light, the maxium speed in the universe, has been a limiting factor that is ever more limiting; I really think that the round trip of SIMD boolean lane-wise test to mask in SIMD register to GPR and back to SIMD will be more significant, our SWAR library already avoids this.  Yet, it is crucial to work with the result of boolean tests in a SIMD way, this is the key to branchless programming techniques, the only way to avoid control dependencies that more than nullify the performance advantage of SIMD parallelism.
```

Comment:
s/worke/work/
ellipsis to period after width.
penultime sentence is seriously run-on.